### DISCUSSION TITLE
Page showing JSON after changing getServerSideProps to point to local API

### DISCUSSION THREAD

**Discussion by fieldr92** [vercel/next.js#19681] [Help] [answered]
Hi there,

We're currently trying to fetch some props before the page loads using getServerSideProps, however since changing the fetch request to point to an application API in the pages/api folder we are getting JSON rendered on the browser instead of HTML.

Docs say the below (see https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering ):-

_"Note: You should not use fetch() to call an API route in your application. Instead, directly import the API route and call its function yourself. You may need to slightly refactor your code for this approach."_

We tried changing our code around but there's no example implementation (where it says you need to slightly refactor) and we're getting the behaviour mentioned above where the page now is rendering JSON, which we think is because we're calling res.end in our api file.

See code we implemented below:-

```
// pages/index.js
import randomAPI from './api/random';

const Home = props => {......};

export async function getServerSideProps({ req, res }) {
  const props = await randomAPI(req, res);

  return { props };
}

export default Home;

-------------------------------------------------------------------------------------------

// pages/api/random.js
export default async function handler(req, res) {
  const response = await fetch('https://run.mocky.io/v3/some-json-output');
  const data = await response.json();

  res.statusCode = 200;
  res.setHeader('Content-Type', 'application/json');
  res.end(JSON.stringify(data));
  return data;
}
```

Please help! Hope that there's enough info here!

Thanks,

Ralph

**Comment by jamesmosier** [comment, marked as answer]
I can't speak to the docs, but I interpret it as "create a function that you can use in your API route and in your `getServerSideProps` function".

For example:

```js
// lib/get-data.js
export default async function GetData() {
  const response = await fetch('https://run.mocky.io/v3/some-json-output');
  const data = await response.json();
  return data;
}
```

and then you can just call that function in your API route as well as `getServerSideProps`.

**Reply by fieldr92** [reply]
Thanks for your response! That works but does open some other questions.

We found it a bit weird that the call in getServerSideProps isn't a proper fetch request (i.e. not a request with a response, just a function call with data returned). In our case, doing the above means that an endpoint is "opened" that doesn't send anything back and the site hangs if you hit that route directly.

Does this mean that getServerSideProps should either hit an app service or an external API? And not a local API?

It seems that way from what we've seen and the interpretation of the docs.

**Reply by jamesmosier** [reply]
> an endpoint is "opened" that doesn't send anything back and the site hangs if you hit that route directly.

I'm not sure I'm understanding. This is what I was thinking:

```js
// lib/get-data.js
export default async function GetData() {
  const response = await fetch('https://run.mocky.io/v3/some-json-output');
  const data = await response.json();
  return data;
}

// pages/index.js
export async function getServerSideProps() {
  const data = await GetData()
  return {
    props: { data }
  }
}

// pages/api/random.js
export default async function handler(req, res) {
  const data = await GetData()
  res.statusCode = 200;
  res.setHeader('Content-Type', 'application/json');
  res.end(JSON.stringify(data));
}
```

**Reply by fieldr92** [reply]
Interesting. That looks like it is what we'll need to do. We had an express middleware before so we were expecting to have to make an endpoint and access with a normal fetch request but I think we need to change how we're thinking about the app for it to work serverless.

Appreciate your help and responses, thanks James!
